os/hal/src/hal_qei.c

```diff
@@ -0,0 +1,214 @@
+/*===========================================================================*/
+
+/**
+ * @brief   QEI Driver initialization.
+ * @note    This function is implicitly invoked by @p halInit(), there is
+ *          no need to explicitly initialize the driver.
+ *
+ * @init
+ */
+void qeiInit(void) {
+
+  qei_lld_init();
+}
+
+/**
+ * @brief   Initializes the standard part of a @p QEIDriver structure.
+ *
+ * @param[out] qeip     pointer to the @p QEIDriver object
+ *
+ * @init
+ */
+void qeiObjectInit(QEIDriver *qeip) {
+
+  qeip->state = QEI_STOP;
+  qeip->last = 0;
+  qeip->config = NULL;
+}
+
+/**
+ * @brief   Configures and activates the QEI peripheral.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ * @param[in] config    pointer to the @p QEIConfig object
+ *
+ * @api
+ */
+void qeiStart(QEIDriver *qeip, const QEIConfig *config) {
+
+  osalDbgCheck((qeip != NULL) && (config != NULL));
+
+  osalSysLock();
+  osalDbgAssert((qeip->state == QEI_STOP) || (qeip->state == QEI_READY),
+                "invalid state");
+  qeip->config = config;
+  qei_lld_start(qeip);
+  qeip->state = QEI_READY;
+  osalSysUnlock();
+}
+
+/**
+ * @brief   Deactivates the QEI peripheral.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ *
+ * @api
+ */
+void qeiStop(QEIDriver *qeip) {
+
+  osalDbgCheck(qeip != NULL);
+
+  osalSysLock();
+  osalDbgAssert((qeip->state == QEI_STOP) || (qeip->state == QEI_READY),
+                "invalid state");
+  qei_lld_stop(qeip);
+  qeip->state = QEI_STOP;
+  osalSysUnlock();
+}
+
+/**
+ * @brief   Enables the quadrature encoder interface.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ *
+ * @api
+ */
+void qeiEnable(QEIDriver *qeip) {
+
+  osalDbgCheck(qeip != NULL);
+
+  osalSysLock();
+  osalDbgAssert(qeip->state == QEI_READY, "invalid state");
+  qei_lld_enable(qeip);
+  qeip->state = QEI_ACTIVE;
+  osalSysUnlock();
+}
+
+/**
+ * @brief   Disables the quadrature encoder interface.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ *
+ * @api
+ */
+void qeiDisable(QEIDriver *qeip) {
+
+  osalDbgCheck(qeip != NULL);
+
+  osalSysLock();
+  osalDbgAssert((qeip->state == QEI_READY) || (qeip->state == QEI_ACTIVE),
+                "invalid state");
+  qei_lld_disable(qeip);
+  qeip->state = QEI_READY;
+  osalSysUnlock();
+}
+
+/**
+ * @brief   Returns the counter value.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ * @return              The current counter value.
+ *
+ * @api
+ */
+qeicnt_t qeiGetCount(QEIDriver *qeip) {
+  qeicnt_t cnt;
+
+  osalSysLock();
+  cnt = qeiGetCountI(qeip);
+  osalSysUnlock();
+
+  return cnt;
+}
+
+/**
+ * @brief   Returns the counter delta from last reading.
+ *
+ * @param[in] qeip      pointer to the @p QEIDriver object
+ * @return              The delta from last read.
+ *
+ * @api
+ */
+qeidelta_t qeiUpdate(QEIDriver *qeip) {
+  qeicnt_t diff;
+
+  osalSysLock();
+  diff = qeiUpdateI(qeip);

```

Comment:
Here you store value in unsigned variable and return them as signed. Is it correct?